src/bounded_loop.rs

```diff
@@ -0,0 +1,156 @@
+use std::fmt;
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::env;
+use std::collections::HashMap;
+use std::fs::OpenOptions;
+use std::io::Write;
+use std::io::{BufReader, BufRead, Read};
+use std::fs::File;
+
+lazy_static!{
+    static ref MUTAGEN_LOOP_COUNT: HashMap<LoopId, usize> = {
+        let mut h = HashMap::new();
+
+        File::open("target/mutagen/loops.txt").map(|mut file| {
+            let reader = BufReader::new(file);
+
+            for l in reader.lines() {
+                let line = l.unwrap_or(String::from(""));
+                let splits: Vec<&str> = line.split(",").collect();
+
+                if splits.len() != 2 {
+                    continue;
+                }
+
+                let lid: usize = splits[0].parse().unwrap_or(0usize);
+                let bound: usize = splits[1].parse().unwrap_or(0usize);
+
+                h.insert(LoopId::new(lid), bound);
+            }
+        });
+
+        h
+    };
+}
+
+#[derive(Clone, Eq, Hash, PartialEq)]
+pub struct LoopId(usize);
+
+impl LoopId {
+    pub fn new(id: usize) -> Self {
+        LoopId(id)
+    }
+
+    pub fn next(&self) -> Self {
+        LoopId(self.0 + 1)
+    }
+
+    pub fn id(&self) -> usize {
+        self.0
+    }
+}
+
+impl fmt::Display for LoopId {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
+/// LoopStep is used to record or impose a bounded limit to a specific loop, which is identified by an id.
+pub trait LoopStep {
+    fn step(&mut self);
+}
+
+/// LoopCount tracks and report the maximum bound for this specific loop.
+pub struct LoopCount<'a> {
+    /// identifies a specific loop
+    id: LoopId,
+    /// count keeps the track of the current execution of this loop
+    count: usize,
+    /// reference to a static allocated AtomicUsize which keeps track of
+    atomic: &'a AtomicUsize,
+}
+
+impl<'a> LoopStep for LoopCount<'a> {
+    /// step is called on each iteration of a loop
+    fn step(&mut self) {
+        self.count = self.count.saturating_add(1);
+    }
+}
+
+impl<'a> Drop for LoopCount<'a> {
+    fn drop(&mut self) {
+        // TODO: llogiq suggested to use atomic max as the following code is not accurate due to TOCTOU
```

Comment:
If I'm not wrong, on some place you talked about using an atomic function which saves the max, but I didn't found it on the documentation. Maybe I'm not looking on the right place or my memory fails.

Reply:
`AtomicUsize::fetch_max`. I added it a few weeks ago, and it's unstable, so it's only in the nightly docs.